src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -295,16 +303,14 @@ internal void Advance(int bytesWritten)
             {
                 Debug.Assert(_writingHead.Next == null);
 
-                Memory<byte> buffer = _writingHead.AvailableMemory;
-
-                if (_writingHead.End > buffer.Length - bytesWritten)
+                if (bytesWritten > _writingMemory.Length)
                 {
                     ThrowHelper.ThrowInvalidOperationException_AdvancingPastBufferSize();
                 }
 
-                // if bytesWritten is zero, these do nothing
-                _writingHead.End += bytesWritten;
                 _currentWriteLength += bytesWritten;
+                _buffered += bytesWritten;
+                _writingMemory = _writingMemory.Slice(bytesWritten);
```

Comment:
Can you help me understand why this change helped. Seems like Slice just moved out from .End setter to here.

Reply:
Slice in `.End` doesn't add value to the writer as its a `ReadOnlyMemory` so doing that it still needs to be resliced on `GetMemory`

This now Slices the `Memory` then so its always read for the writer, then Slices via `.End` for the reader on `.Commit`

Reply:
There are 2 sets of slicing that need to happen:
 1. Changing the memory for the reader - `AvaliableMemory.Slice(0, bytesWritten)`
 2. Changing the memory for the writer - `AvailableMemory.Slice(bytesWritten)`

Today every call to GetMemory/GetSpan does a slice to give expose the writable space and Advance updates the memory available to the reader. It means when you GetMemory/GetSpan and Advance we update both pieces of memory all of the time. This change makes it so that we update the appropriate side when necessary (on Flush and when we run out of space in the existing segment). Other performance comes from the fact that we're not going through the segment every time to get the writable memory since it's just a field.

Reply:
Why don't we just have 2 properties on Segment then? One for writing one for reading pointing to the same memory?

Reply:
> Other performance comes from the fact that we're not going through the segment every time to get the writable memory since it's just a field.

Is this proven? How expensive is the indirection?

Reply:
> Why don't we just have 2 properties on Segment then?

Then you'll bloat the segment size of which there are many for the writing head which there is one?

Reply:
Specifically by 16 bytes
```
Type layout for 'Memory`1'
Size: 16 bytes. Paddings: 0 bytes (%0 of empty space)
|=================================|
|   0-7: Object _object (8 bytes) |
|---------------------------------|
|  8-11: Int32 _index (4 bytes)   |
|---------------------------------|
| 12-15: Int32 _length (4 bytes)  |
|=================================|
```

Reply:
Now I understand, thank you all.

Reply:
Motivation behind this change is @davidfowl thought doing this upstream was an abomination https://github.com/aspnet/AspNetCore/pull/7432 😄